sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -112,6 +125,28 @@ private static DoFnSignature parseSignature(Class<? extends DoFn<?, ?>> fnClass)
         findAnnotatedMethod(errors, DoFn.GetRestrictionCoder.class, fnClass, false);
     Method newTrackerMethod = findAnnotatedMethod(errors, DoFn.NewTracker.class, fnClass, false);
 
+    Collection<Method> onTimerMethods =
+        declaredMethodsWithAnnotation(DoFn.OnTimer.class, fnClass, DoFn.class);
+    HashMap<String, DoFnSignature.OnTimerMethod> onTimerMethodMap =
+            Maps.newHashMapWithExpectedSize(onTimerMethods.size());
+    for (Method onTimerMethod : onTimerMethods) {
+      String id = onTimerMethod.getAnnotation(DoFn.OnTimer.class).value();
+        errors.checkArgument(timerDeclarations.containsKey(id),
+            "Callback %s is for for undeclared timer %s", onTimerMethod, id);
+      onTimerMethodMap.put(id, OnTimerMethod.create(onTimerMethod, id, Collections.EMPTY_LIST));
+    }
+    builder.setOnTimerMethods(onTimerMethodMap);
+
+    // Check the converse - that all timers have a callback. This could be relaxed to only
+    // those timers used in methods, once method parameter lists support timers.
```

Comment:
I'd like to submit the annotations in pieces and enable them later, to keep code reviews ~500 LOC or certainly <1000.

What I started with was javadoc only for the annotations that were currently allowed to be written, which naturally bothered reviewers since it was necessarily dead code. So I fleshed it out to describe the full API, which perhaps was a mistake. The alternative I see is to make it very terse and clearly unsupported. Would you prefer that? I am quite happy to do that until they can all be written down together. (one PR from now, in progress)